apps/prairielearn/elements/pl-file-editor/pl-file-editor.py

```diff
@@ -129,12 +160,21 @@ def render(element_html, data):
             directory = os.path.join(data["options"]["question_path"], directory)
         file_path = os.path.join(directory, source_file_name)
         text_display = open(file_path).read()
+
+        #check if the file provided should be a static editor
+        if "<static>" in text_display:
```

Comment:
I like the idea of being able to make some lines static, but I would strongly prefer if these could be specified by line number (as how the `pl-code` element specifies lines to highlight, see [here](https://github.com/PrairieLearn/PrairieLearn/blob/master/exampleCourse/questions/element/code/question.html) for an example), since the current implementation has the static tags appear in source files.

Reply:
I suggested the same thing (line numbers in an attribute) during an initial chat in office hours and said we'd continue the conversation once a PR was opened. I'm still leaning towards recommending that as the direction we take this in, but it'd be nice to have input from other reviewers as well.

Reply:
I think line numbers are probably the best option, but they should be clearly set up to count the lines in the starting text. In other words, after a submission that adds new lines, the fixed lines may shift around (e.g., line 4 above could be 5 or 6 in the submitted answer).

Reply:
An alternative option is to have some customizable annotation on editable (or static) lines. For example:
```
<pl-file-editor ... editable-line-prefix="@@">
def add(num1, num2):
  sum = 0
@@  // complete this part
  return sum
</pl-file-editor>
```

This is just an idea, feel free to ignore it, but I believe this would make things easier for an instructor to set up. Again, just make sure you have a plan for keeping track of what content is static or editable after a submission.

Reply:
Thanks everyone for your feedback so far. We will get to work on fixing the technical issues you brought up but I'd like to discuss more about how the static/editable sections are specified. Right now it seems like the preference is for the editable (or static) lines to be specified separately in the top level element so as to keep the interior text pure. I would still prefer the ability to specify the editability of a section text in-line for the following reasons
1. It is easier for the text and editability to stay in sync when the underlying question text is edited. I just know I am going to make the mistake where I modify the question text and forget to update the text ranges and then my whole question is off. It is even more annoying if there are multiple sections that need to be updated as they will all cascade into each other.
2. I think in-lining the editability is an easier and more friendly UI. If I have to do line numbers then I have to keep looking at my editor for the line numbers to find the correct ranges. 
    - If the user is isn't sourcing the question text from a file but is instead in-lining the question text within the element the line numbers of the editor will be won't match how many lines into the element they are. They will either have to do some small arithmetic to find the true line numbers or copy and paste the question text into another editor window to get the line numbers. I supposed a work around to this is to use the absolute line numbers in the question.html when the question text is in-lined instead of how far into the element the line numbers are but that would likely exacerbate issue 1.
    - If we want to support intra-line editability, which I could and would use, that is much easier to specify in-line then specifying the character ranges.   
    - When the editability is in-lined, I think it is much easier for a reader and myself to know what is/is not editable.

Reply:
Looking at this again, I'm less dead-set about using concrete line numbers for something like this than other instances where adding extra syntax has been proposed on top of existing languages (there's no guarantee the starter code is even valid code, which is fine if students are expected to fill something in). I'm not sure I'm a huge fan of the syntax and I'm wondering if there are other examples of technologies that do something like this that we could emulate (or even better, grab a python / js library and have it do the work for us).

I would however prefer enabling editable areas via an attribute rather than the automatic detection that is done now. But @mfbutner I agree that having something directly in the starter code that specifies what's read-only and what isn't makes it easier to keep track of everything for the user.

Reply:
My concern with inlining these is how they could interact with elements of the language itself. So, for example, using `<static>` could produce some issues if the code being edited is HTML. I'm not sure if I have a solution for it, if you can suggest something I'm happy to hear.

Reply:
We aren't adamant that the way that the static/editable sections be specified via a nested html tag. We simply went that route because we had seen other pl-elements use nested elements when some interior items needed additional information. More what I'm advocating for is a way to specify the static/editable sections in-line as I feel that is the easiest, natural, and readable way to do so. 

For 
> where adding extra syntax has been proposed on top of existing languages (there's no guarantee the starter code is even valid code, which is fine if students are expected to fill something in).

and

>My concern with inlining these is how they could interact with elements of the language itself. So, for example, using <static> could produce some issues if the code being edited is HTML.

I think the common concern, and please correct me if I'm wrong, is that the in-line specification of the editability might interfere with the interpretation/parsing of the text the user wants rendered. Using jonatan's example someone might have a pl-file-editor where they want to display some html and they have a tag named static/editable that they want to appear but we now accidentally interpret as specifying the editability and so strip  it out of what is rendered in the question. 

>  I'm not sure I'm a huge fan of the syntax and I'm wondering if there are other examples of technologies that do something like this that we could emulate (or even better, grab a python / js library and have it do the work for us).

The closest thing that addresses these issues that I've used in the past are [filters from jinja2](https://jinja.palletsprojects.com/en/3.1.x/templates/#id11). Using jinja2 the question would look like

```
<pl-file-editor>
{% filter editable %}
  You CAN edit this text
{% endfilter %}

{% filter static %}
  You can  NOT edit this text
{% endfilter %}


{% filter editable %}
  You CAN  ALSO edit this text
{% endfilter %}

</pl-file-editor>
```

Filters can accept arguments as well so if we wanted to limit the maximum number of lines of an editable section to 5 it would look  something like like

```
<pl-file-editor>
{% filter editable(max_height=5) %}
  You CAN edit this text
{% endfilter %}
```




 If we didn't want to bring in a new template parser for this element, Mustache has something [very similar to this called lambdas](https://mustache.github.io/mustache.5.html). Given that each element is responsible for rendering its contents, I think it would be possible to piggy back onto this. So now a question's text would look like
```
<pl-file-editor>
{{#editable}}
  You CAN edit this text
{{/editable}}

{{#static}}
  You can  NOT edit this text
{{/static}}

{{#editable}}
  You CAN  ALSO edit this text
{{/editable}}
</pl-file-editor>
```

Unlike jinja's filters, mustache's lambdas don't have a builtin way to pass more information than the text that should be processed. If we don't want to use jinja for this element, we could take the harder path and hack parameters into the lambda. I've been looking online and it seems like there are some workarounds where you specify the parameters in the text and then extract them from it. As a very rough first idea it might look like if we wanted to set the maximum number of lines to 5

```
<pl-file-editor>
{{#editable}}
  {{max_height=5}}
  You CAN edit this text
{{/editable}}
```

The exact syntax we would want to use for specifying those extra parameters is completely left up to us so please don't consider this a final draft. 


Unfortunately, even with this I feel like an in-line specification could always lead to some conflict with the underlying text. The goal would be to minimize the conflict while at the same time allowing an escape mechanism if one does occur. 
I look forward to your suggestions on this alternative. It would also help to hear how you were creating a question that used these features what you would like it to look like.

Reply:
I don't think mustache would be a good idea, particularly if the editor text is added inside question.html itself. The issue is that mustache is parsed in question.html before elements are, so the mustache code would likely not reach the editor.

I don't have a strong opinion for or against jinja, though I'd be hesitant to add a new framework to the mix. I would be fine with keeping something like `<editable>` or equivalent notation, as long as (1) this is done with substring search only, not with html parsing, and (2) there is a way to change this for code where this would be a problem. I don't know what the appropriate solution is, maybe a parameter that indicates what the start/end keywords are, with a default like `<editable>`? @nwalters512 or @mwest1066 may have a stronger (and more authoritative) opinion here.

I should point out that you don't need both `<editable>` and `<static>`, you could have only one of them and infer the other from what's left.